// captcha-canvas/dist/security/SecurityPresets.js
"use strict";
/**
 * Security configuration presets for CAPTCHA generation.
 *
 * This module provides predefined security levels that configure character set complexity,
 * text length requirements, entropy levels, and validation strictness to balance
 * security against usability for different use cases.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SECURITY_PRESETS = exports.MAXIMUM_SECURITY = exports.HIGH_SECURITY = exports.MEDIUM_SECURITY = exports.LOW_SECURITY = exports.SecurityLevel = void 0;
exports.getSecurityConfig = getSecurityConfig;
exports.validateSecurityConfig = validateSecurityConfig;
exports.getSecurityInfo = getSecurityInfo;
exports.createCustomSecurityConfig = createCustomSecurityConfig;
exports.recommendSecurityLevel = recommendSecurityLevel;
const CryptoUtils_1 = require("./CryptoUtils");
/**
 * Security level enumeration.
 */
var SecurityLevel;
(function (SecurityLevel) {
    SecurityLevel["LOW"] = "LOW";
    SecurityLevel["MEDIUM"] = "MEDIUM";
    SecurityLevel["HIGH"] = "HIGH";
    SecurityLevel["MAXIMUM"] = "MAXIMUM";
})(SecurityLevel || (exports.SecurityLevel = SecurityLevel = {}));
/**
 * LOW security preset - Basic protection suitable for low-risk applications.
 *
 * Provides minimal security with high usability. Suitable for comment forms,
 * newsletter signups, or other low-risk scenarios where user experience
 * is prioritized over security.
 *
 * @example
 * ```typescript
 * const captcha = new CaptchaGenerator()
 *   .setSecurityLevel(SecurityLevel.LOW);
 * ```
 */
exports.LOW_SECURITY = {
    level: SecurityLevel.LOW,
    characterSet: CryptoUtils_1.CHARACTER_SETS.HEXADECIMAL,
    textLength: {
        min: 4,
        max: 6,
        default: 5
    },
    entropy: {
        minimum: 16, // 4 chars * 4 bits = 16 bits minimum
        recommended: 20 // 5 chars * 4 bits = 20 bits recommended
    },
    validation: {
        requireCrypto: false, // Allow Math.random fallback for compatibility
        enableSanitization: true,
        enableNormalization: false,
        maxAttempts: 3
    },
    visual: {
        minRotation: 0,
        maxRotation: 10,
        enableSkew: false,
        minOpacity: 0.8,
        decoyMultiplier: 1.0
    },
    rateLimiting: {
        requestsPerMinute: 60,
        sessionTimeoutMinutes: 30
    }
};
/**
 * MEDIUM security preset - Balanced security and usability.
 *
 * Provides good security with reasonable usability. Suitable for most
 * web applications, user registration forms, and general-purpose CAPTCHAs.
 * This is the recommended default for most applications.
 *
 * @example
 * ```typescript
 * const captcha = new CaptchaGenerator()
 *   .setSecurityLevel(SecurityLevel.MEDIUM);
 * ```
 */
exports.MEDIUM_SECURITY = {
    level: SecurityLevel.MEDIUM,
    characterSet: CryptoUtils_1.CHARACTER_SETS.ALPHANUMERIC,
    textLength: {
        min: 5,
        max: 7,
        default: 6
    },
    entropy: {
        minimum: 25, // ~5 chars * 5.17 bits ≈ 25 bits minimum
        recommended: 31 // ~6 chars * 5.17 bits ≈ 31 bits recommended
    },
    validation: {
        requireCrypto: true, // Require cryptographic security
        enableSanitization: true,
        enableNormalization: true,
        maxAttempts: 5
    },
    visual: {
        minRotation: 5,
        maxRotation: 15,
        enableSkew: true,
        minOpacity: 0.7,
        decoyMultiplier: 1.5
    },
    rateLimiting: {
        requestsPerMinute: 30,
        sessionTimeoutMinutes: 15
    }
};
/**
 * HIGH security preset - Strong security for sensitive applications.
 *
 * Provides high security with moderate usability impact. Suitable for
 * financial applications, admin panels, password reset forms, and other
 * security-sensitive scenarios.
 *
 * @example
 * ```typescript
 * const captcha = new CaptchaGenerator()
 *   .setSecurityLevel(SecurityLevel.HIGH);
 * ```
 */
exports.HIGH_SECURITY = {
    level: SecurityLevel.HIGH,
    characterSet: CryptoUtils_1.CHARACTER_SETS.ALPHANUMERIC_MIXED_CASE,
    textLength: {
        min: 6,
        max: 8,
        default: 7
    },
    entropy: {
        minimum: 35, // ~6 chars * 5.95 bits ≈ 35 bits minimum
        recommended: 42 // ~7 chars * 5.95 bits ≈ 42 bits recommended
    },
    validation: {
        requireCrypto: true,
        enableSanitization: true,
        enableNormalization: true,
        maxAttempts: 3
    },
    visual: {
        minRotation: 10,
        maxRotation: 25,
        enableSkew: true,
        minOpacity: 0.6,
        decoyMultiplier: 2.0
    },
    rateLimiting: {
        requestsPerMinute: 15,
        sessionTimeoutMinutes: 10
    }
};
/**
 * MAXIMUM security preset - Maximum security for critical applications.
 *
 * Provides maximum security with significant usability impact. Suitable for
 * highly sensitive applications, security research, or scenarios where
 * security is paramount over user experience.
 *
 * @example
 * ```typescript
 * const captcha = new CaptchaGenerator()
 *   .setSecurityLevel(SecurityLevel.MAXIMUM);
 * ```
 */
exports.MAXIMUM_SECURITY = {
    level: SecurityLevel.MAXIMUM,
    characterSet: CryptoUtils_1.CHARACTER_SETS.ALPHANUMERIC_SYMBOLS,
    textLength: {
        min: 8,
        max: 12,
        default: 10
    },
    entropy: {
        minimum: 48, // ~8 chars * 6.17 bits ≈ 48 bits minimum
        recommended: 62 // ~10 chars * 6.17 bits ≈ 62 bits recommended
    },
    validation: {
        requireCrypto: true,
        enableSanitization: true,
        enableNormalization: true,
        maxAttempts: 2
    },
    visual: {
        minRotation: 15,
        maxRotation: 35,
        enableSkew: true,
        minOpacity: 0.5,
        decoyMultiplier: 3.0
    },
    rateLimiting: {
        requestsPerMinute: 5,
        sessionTimeoutMinutes: 5
    }
};
/**
 * Security preset lookup table.
 */
exports.SECURITY_PRESETS = {
    [SecurityLevel.LOW]: exports.LOW_SECURITY,
    [SecurityLevel.MEDIUM]: exports.MEDIUM_SECURITY,
    [SecurityLevel.HIGH]: exports.HIGH_SECURITY,
    [SecurityLevel.MAXIMUM]: exports.MAXIMUM_SECURITY
};
/**
 * Gets the security configuration for the specified level.
 *
 * @param level - Security level to retrieve
 * @returns Security configuration object
 * @throws {Error} When security level is invalid
 *
 * @example
 * ```typescript
 * const config = getSecurityConfig(SecurityLevel.HIGH);
 * console.log(`Using ${config.characterSet.name} with ${config.textLength.default} characters`);
 * ```
 */
function getSecurityConfig(level) {
    const config = exports.SECURITY_PRESETS[level];
    if (!config) {
        throw new Error(`Invalid security level: ${level}. Valid levels: ${Object.keys(SecurityLevel).join(', ')}`);
    }
    return { ...config }; // Return a copy to prevent modification
}
/**
 * Validates that a security configuration meets minimum requirements.
 *
 * @param config - Security configuration to validate
 * @throws {Error} When configuration is invalid
 *
 * @example
 * ```typescript
 * const customConfig = { ...MEDIUM_SECURITY, textLength: { min: 2, max: 4, default: 3 } };
 * validateSecurityConfig(customConfig); // May throw if entropy is too low
 * ```
 */
function validateSecurityConfig(config) {
    // Validate text length constraints
    if (config.textLength.min < 1) {
        throw new Error('Minimum text length must be at least 1');
    }
    if (config.textLength.max < config.textLength.min) {
        throw new Error('Maximum text length must be greater than or equal to minimum');
    }
    if (config.textLength.default < config.textLength.min || config.textLength.default > config.textLength.max) {
        throw new Error('Default text length must be within min/max range');
    }
    // Validate entropy requirements
    if (config.entropy.minimum < 8) {
        throw new Error('Minimum entropy must be at least 8 bits');
    }
    if (config.entropy.recommended < config.entropy.minimum) {
        throw new Error('Recommended entropy must be greater than or equal to minimum');
    }
    // Validate character set
    if (!config.characterSet.charset || config.characterSet.charset.length < 2) {
        throw new Error('Character set must contain at least 2 characters');
    }
    // Validate visual settings
    if (config.visual.minRotation < 0 || config.visual.maxRotation < config.visual.minRotation) {
        throw new Error('Invalid rotation angle settings');
    }
    if (config.visual.minOpacity < 0 || config.visual.minOpacity > 1) {
        throw new Error('Minimum opacity must be between 0 and 1');
    }
    if (config.visual.decoyMultiplier < 0) {
        throw new Error('Decoy multiplier must be non-negative');
    }
    // Validate rate limiting
    if (config.rateLimiting.requestsPerMinute < 1) {
        throw new Error('Requests per minute must be at least 1');
    }
    if (config.rateLimiting.sessionTimeoutMinutes < 1) {
        throw new Error('Session timeout must be at least 1 minute');
    }
}
/**
 * Gets security information and recommendations for a given configuration.
 *
 * @param config - Security configuration to analyze
 * @returns Security analysis object
 *
 * @example
 * ```typescript
 * const analysis = getSecurityInfo(MEDIUM_SECURITY);
 * console.log(`Security Level: ${analysis.level}`);
 * console.log(`Estimated Strength: ${analysis.strength.level}`);
 * console.log(`Recommendations: ${analysis.recommendations.join(', ')}`);
 * ```
 */
function getSecurityInfo(config) {
    const strength = (0, CryptoUtils_1.getCryptographicStrength)(config.textLength.default, config.characterSet);
    const recommendations = [];
    const warnings = [];
    // Analyze entropy
    if (strength.entropyBits < config.entropy.minimum) {
        warnings.push(`Entropy below minimum requirement (${strength.entropyBits.toFixed(1)} < ${config.entropy.minimum} bits)`);
        recommendations.push('Increase text length or use a larger character set');
    }
    else if (strength.entropyBits < config.entropy.recommended) {
        recommendations.push(`Consider increasing entropy to recommended level (${config.entropy.recommended} bits)`);
    }
    // Analyze crypto requirements
    if (!config.validation.requireCrypto) {
        warnings.push('Math.random() fallback allowed - reduces security');
        recommendations.push('Enable requireCrypto for better security');
    }
    // Analyze visual security
    if (!config.visual.enableSkew) {
        recommendations.push('Enable skewing for better OCR resistance');
    }
    if (config.visual.maxRotation < 10) {
        recommendations.push('Consider increasing rotation range for better security');
    }
    // Analyze rate limiting
    if (config.rateLimiting.requestsPerMinute > 60) {
        warnings.push('High rate limit may allow brute force attacks');
        recommendations.push('Consider reducing requests per minute');
    }
    return {
        level: config.level,
        strength,
        recommendations,
        warnings
    };
}
/**
 * Creates a custom security configuration based on a preset with overrides.
 *
 * @param baseLevel - Base security level to start from
 * @param overrides - Partial configuration to override base settings
 * @returns Custom security configuration
 *
 * @example
 * ```typescript
 * const customConfig = createCustomSecurityConfig(SecurityLevel.MEDIUM, {
 *   textLength: { min: 8, max: 10, default: 9 },
 *   characterSet: CHARACTER_SETS.ALPHANUMERIC_SYMBOLS
 * });
 * ```
 */
function createCustomSecurityConfig(baseLevel, overrides) {
    const baseConfig = getSecurityConfig(baseLevel);
    const customConfig = {
        ...baseConfig,
        ...overrides,
        // Deep merge nested objects
        textLength: { ...baseConfig.textLength, ...overrides.textLength },
        entropy: { ...baseConfig.entropy, ...overrides.entropy },
        validation: { ...baseConfig.validation, ...overrides.validation },
        visual: { ...baseConfig.visual, ...overrides.visual },
        rateLimiting: { ...baseConfig.rateLimiting, ...overrides.rateLimiting }
    };
    // Validate the custom configuration
    validateSecurityConfig(customConfig);
    return customConfig;
}
/**
 * Recommends an appropriate security level based on application requirements.
 *
 * @param requirements - Application security requirements
 * @returns Recommended security level and reasoning
 *
 * @example
 * ```typescript
 * const recommendation = recommendSecurityLevel({
 *   isFinancial: true,
 *   hasPersonalData: true,
 *   expectedVolume: 'medium',
 *   userExperiencePriority: 'balanced'
 * });
 * console.log(`Recommended: ${recommendation.level} - ${recommendation.reasoning}`);
 * ```
 */
function recommendSecurityLevel(requirements) {
    let score = 0;
    const reasons = [];
    // Increase score based on risk factors
    if (requirements.isFinancial) {
        score += 3;
        reasons.push('financial application');
    }
    if (requirements.hasPersonalData) {
        score += 2;
        reasons.push('handles personal data');
    }
    if (requirements.isAdminPanel) {
        score += 2;
        reasons.push('administrative interface');
    }
    // Adjust based on threat level
    switch (requirements.threatLevel) {
        case 'critical':
            score += 4;
            reasons.push('critical threat level');
            break;
        case 'high':
            score += 3;
            reasons.push('high threat level');
            break;
        case 'medium':
            score += 1;
            reasons.push('medium threat level');
            break;
    }
    // Adjust based on user experience priority
    switch (requirements.userExperiencePriority) {
        case 'security':
            score += 2;
            reasons.push('security prioritized over UX');
            break;
        case 'usability':
            score -= 2;
            reasons.push('usability prioritized over security');
            break;
    }
    // Adjust based on expected volume (higher volume = higher risk)
    switch (requirements.expectedVolume) {
        case 'high':
            score += 1;
            reasons.push('high traffic volume');
            break;
        case 'low':
            score -= 1;
            reasons.push('low traffic volume');
            break;
    }
    // Determine security level based on score
    let level;
    let alternatives = [];
    if (score >= 8) {
        level = SecurityLevel.MAXIMUM;
        alternatives = [SecurityLevel.HIGH];
    }
    else if (score >= 5) {
        level = SecurityLevel.HIGH;
        alternatives = [SecurityLevel.MAXIMUM, SecurityLevel.MEDIUM];
    }
    else if (score >= 2) {
        level = SecurityLevel.MEDIUM;
        alternatives = [SecurityLevel.HIGH, SecurityLevel.LOW];
    }
    else {
        level = SecurityLevel.LOW;
        alternatives = [SecurityLevel.MEDIUM];
    }
    const reasoning = reasons.length > 0
        ? `Recommended due to: ${reasons.join(', ')}`
        : 'Default recommendation for general use';
    return { level, reasoning, alternatives };
}
