// captcha-canvas/dist/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultDecoyOptions = exports.defaultTraceOptions = exports.defaultCaptchaOptions = void 0;
/**
 * Default configuration options for captcha text.
 *
 * These values are used when no custom options are provided. They represent
 * a balanced configuration that provides good security while maintaining readability.
 *
 * @example
 * ```typescript
 * // These are the default values used internally
 * const defaults = {
 *   characters: 6,        // 6 random characters
 *   size: 40,            // 40px font size
 *   font: 'Sans',        // Sans-serif font
 *   skew: true,          // Enable skewing
 *   colors: [],          // No color array (uses single color)
 *   rotate: 5,           // Max 5 degrees rotation
 *   color: '#32cf7e',    // Green color
 *   opacity: 0.8         // 80% opacity
 * };
 * ```
 */
exports.defaultCaptchaOptions = {
    characters: 6,
    size: 40,
    font: 'Sans',
    skew: true,
    colors: [],
    rotate: 5,
    color: '#32cf7e',
    opacity: 0.8,
};
/**
 * Default configuration options for trace lines.
 *
 * These values provide a good balance between security enhancement and visual appeal.
 * The trace lines are fully opaque and use the same color as the default text.
 *
 * @example
 * ```typescript
 * // Default trace line configuration
 * const defaults = {
 *   size: 3,             // 3px line width
 *   color: '#32cf7e',    // Same green as default text
 *   opacity: 1           // Fully opaque
 * };
 * ```
 */
exports.defaultTraceOptions = {
    size: 3,
    color: '#32cf7e',
    opacity: 1,
};
/**
 * Default configuration options for decoy characters.
 *
 * These values create subtle background noise that confuses OCR systems without
 * significantly impacting human readability. The decoy characters are smaller
 * and use a neutral gray color.
 *
 * @example
 * ```typescript
 * // Default decoy character configuration
 * const defaults = {
 *   color: '#646566',    // Neutral gray color
 *   font: 'Sans',        // Same font as main text
 *   size: 20,            // Smaller than default text (40px)
 *   opacity: 0.8         // 80% opacity
 * };
 * ```
 */
exports.defaultDecoyOptions = {
    color: '#646566',
    font: 'Sans',
    size: 20,
    opacity: 0.8,
};
