// captcha-canvas/dist/errors.js
"use strict";
/**
 * Custom error types for captcha-canvas to provide actionable and typed error reporting.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CaptchaSecurityError = exports.CaptchaGenerationError = exports.CaptchaValidationError = exports.CaptchaBaseError = void 0;
class CaptchaBaseError extends Error {
    constructor(message, code, details) {
        super(message);
        this.name = this.constructor.name;
        this.code = code;
        this.details = details;
        // Maintains proper stack trace for where our error was thrown (only on V8)
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.CaptchaBaseError = CaptchaBaseError;
/**
 * Thrown when configuration/parameter validation fails.
 */
class CaptchaValidationError extends CaptchaBaseError {
    constructor(message, details) {
        super(message, 'VALIDATION_ERROR', details);
    }
}
exports.CaptchaValidationError = CaptchaValidationError;
/**
 * Thrown when rendering/generation fails (e.g., canvas/image errors).
 */
class CaptchaGenerationError extends CaptchaBaseError {
    constructor(message, details) {
        super(message, 'GENERATION_ERROR', details);
    }
}
exports.CaptchaGenerationError = CaptchaGenerationError;
/**
 * Thrown when security-related constraints are violated.
 */
class CaptchaSecurityError extends CaptchaBaseError {
    constructor(message, details) {
        super(message, 'SECURITY_ERROR', details);
    }
}
exports.CaptchaSecurityError = CaptchaSecurityError;
