// discord-arts/dist/constants/assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FONTS = void 0;
exports.FONTS = {
    HELVETICA: {
        regular: 'Helvetica',
        bold: 'Helvetica Bold',
        paths: {
            regular: 'Helvetica.ttf',
            bold: 'HelveticaBold.ttf',
        },
    },
    NOTO_SANS: {
        regular: 'Noto Sans',
        bold: 'Noto Sans Bold',
        paths: {
            regular: 'NotoSans_SemiCondensed-Regular.ttf',
            bold: 'NotoSans_SemiCondensed-Bold.ttf',
        },
    },
    ROBOTO: {
        regular: 'Roboto',
        bold: 'Roboto Bold',
        paths: {
            regular: 'RobotoCondensed-Regular.ttf',
            bold: 'RobotoCondensed-Bold.ttf',
        },
    },
    OPEN_SANS: {
        regular: 'Open Sans',
        bold: 'Open Sans Bold',
        paths: {
            regular: 'OpenSans_SemiCondensed-Regular.ttf',
            bold: 'OpenSans_SemiCondensed-Bold.ttf',
        },
    },
    MONTSERRAT: {
        regular: 'Montserrat',
        bold: 'Montserrat Bold',
        paths: {
            regular: 'Montserrat-Regular.ttf',
            bold: 'Montserrat-Bold.ttf',
        },
    },
};
